ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c

```diff
@@ -0,0 +1,350 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ */
+/**
+ * @file	tensordec-directvideo.c
+ * @date	19 Oct 2019
+ * @brief	NNStreamer tensor-decoder subplugin, "image segment",
+ *              which detects objects and paints their regions.
+ *
+ * @see		https://github.com/nnsuite/nnstreamer
+ * @author	Jihoon Lee <ulla4571@gmail.com>
+ *          niklasjang <niklasjang@gmail.com>
+ * @bug		No known bugs except for NYI items
+ *
+ * option1: Decoder mode of image segmentation
+ *          Available : tflite-deeplab
+ * option2: Video Output Dimension (WIDTH:HEIGHT)
+ *          This depends on option1 values
+ * option3: Video Input Dimension (WIDHT: HEIGHT)
+ *          This depends on option1 values
+ */
+
+#include <string.h>
+#include <glib.h>
+#include <gst/video/video-format.h>
+#include <nnstreamer_plugin_api_decoder.h>
+#include <nnstreamer_plugin_api.h>
+
+void init_is (void) __attribute__ ((constructor));
+void fini_is (void) __attribute__ ((destructor));
+
+#define TFLITE_DEEPLAB_TOTAL_LABELS    21
+#define TFLITE_DEEPLAB_IMAGE_WIDTH     257
+#define TFLITE_DEEPLAB_IMAGE_HEIGHT    257
+
+/**
+ * @brief There can be different schemes for image segmentation.
+ */
+typedef enum
+{
+  TFLITE_DEEPLAB_IMAGE_SEGMENT = 0,
+  UNKNOWN_IMAGE_SEGMENT,
+} image_segment_modes;
+
+/**
+ * @brief List of image-segmentation decoding schemes in string
+ */
+static const char *is_modes[] = {
+  [TFLITE_DEEPLAB_IMAGE_SEGMENT] = "tflite-deeplab",
+  NULL,
+};
+
+/**
+ * @brief Data structure for image segmentation info.
+ */
+typedef struct
+{
+  image_segment_modes mode; /**< The image segmentation decoding mode */
+  guint **segment_map;  /**< The image segmentated map */
+
+  /*  From option2 */
+  guint o_width; /**< Output video Width */
+  guint o_height; /**< Output video Height */
+
+  /*  From option3 */
+  guint i_width; /**< Input video Width */
+  guint i_height; /**< Input video Height */
+} image_segments;
+
+/** @brief Initialize image_segments per mode */
+static int
+_init_modes (image_segments * idata)
+{
+  if (idata->mode == TFLITE_DEEPLAB_IMAGE_SEGMENT) {
+    int i;
+
+    idata->i_width = TFLITE_DEEPLAB_IMAGE_WIDTH;
+    idata->i_height = TFLITE_DEEPLAB_IMAGE_HEIGHT;
+    idata->o_width = TFLITE_DEEPLAB_IMAGE_WIDTH;
+    idata->o_height = TFLITE_DEEPLAB_IMAGE_HEIGHT;
+    idata->segment_map = g_new0 (guint *, idata->i_height);
+    for (i = 0; i < idata->i_height; i++) {
+      idata->segment_map[i] = g_new0 (guint, idata->i_width);
+    }
+  }
+
+  return TRUE;
+}
+
+/** @brief tensordec-plugin's GstTensorDecoderDef callback */
+static int
+is_init (void **pdata)
+{
+  image_segments *idata;
+  *pdata = g_new0 (image_segments, 1);
+
+  idata = *pdata;
+  idata->mode = UNKNOWN_IMAGE_SEGMENT;
+  idata->i_width = 0;
+  idata->i_height = 0;
+  idata->o_width = 0;
```

Comment:
Do we really need o_width and o_height?

Reply:
May be not. I'll remove it

Reply:
Do we need i_width and i_height by option?